### PR TITLE
Improve record type errors

### PR DESCRIPTION
**Pull request by henridf**
I will add some tests, but wanted to get some 👀 on this first.

### CODE REVIEW

File: pkg/zio/zjsonio/reader.go [hunk 1]
```diff
@@ -193,7 +193,7 @@ func decodeContainer(builder *zval.Builder, typ zeek.Type, body []interface{}) e
 		}
 		if columns != nil {
 			if k >= len(columns) {
-				return zng.ErrTypeMismatch
+				return &zng.RecordTypeError{Name: "<record>", Type: typ.String(), Err: zng.ErrExtraField}
```
Comment (mccanne) [inline pkg/zio/zjsonio/reader.go]: It seems like a look syntactic sugar here would be nice... maybe a helper function to create these error objects?

File: pkg/zng/zng.go [hunk 2]
```diff
@@ -13,12 +13,21 @@ var (
 	ErrDescriptorInvalid = errors.New("zng descriptor out of range")
 	ErrBadValue          = errors.New("malformed zng value")
 	ErrBadFormat         = errors.New("malformed zng record")
-	ErrTypeMismatch      = errors.New("zng type/value mismatch")
+	ErrAccessWrongType   = errors.New("typed field access on wrong type")
```
Comment (mccanne) [inline pkg/zng/zng.go]: This doesn't seem like a better description but maybe I'm missing something.  Where are these errors cropping up?

File: reducer/field/count.go [hunk 3]
```diff
@@ -24,7 +23,7 @@ func (c *Count) Consume(v zeek.Value) error {
 	var i zeek.Int
 	//XXX need CoerceToCount?
 	if !zeek.CoerceToInt(v, &i) {
-		return zng.ErrTypeMismatch
+		return errCoerce
 	}
```
Comment (mccanne) [inline reducer/field/count.go]: Seems like errCoerce should be exportable.  And I would again question why isn't this a type mismatch?

Comment (mccanne) [reply reducer/field/count.go]: Is this error even exported?  It's a way for the reducers to track field values that were encountered that cannot be reduced according to the function applied.  It seems like ErrTypeMismatch is the right error here.

**Comment by henridf** [comment]
@mccanne you're right about ErrTypeMismatch. Since it was overloaded I started off moving some uses to their own errors, in order to make room for the record structural type errors, and then realized that those ones deserved a different name (I ended up with RecordTypeError). I'll clean this up before merging.

File: pkg/zng/zng.go (hunk 2 above)
Comment (henridf) [reply pkg/zng/zng.go]: reverted to ErrTypeMismatch.

File: reducer/field/count.go (hunk 3 above)
Comment (henridf) [reply reducer/field/count.go]: Reverted. Now I've worked on this a little, I think what matters most is properly wrapping errors to give context on where an error happened. This PR does this for record typechecks; others remain to be done in follow-on PRs.

File: pkg/zio/zjsonio/reader.go (hunk 1 above)
Comment (henridf) [reply pkg/zio/zjsonio/reader.go]: `newRecordTypeError("record", typ.String(), zng.ErrExtraField)` is a few less characters but also adds one hop of indirection for the reader to see what is being returned.

**Comment by henridf** [comment]
(skipped tests for now... this feels still a bit fresh)

File: pkg/zng/record.go [hunk 4]
```diff
@@ -204,95 +204,122 @@ func (r *Record) ZeekStrings(precision int, utf8 bool) ([]string, bool, error) {
...
 	return checkRecord(r.Descriptor.Type, r.Raw)
 }
 
-// check a vector whose inner type is a record
-func checkVector(typ *zeek.TypeVector, body zval.Encoding) bool {
+var (
+	ErrMissingField = errors.New("record missing a field")
+	ErrExtraField   = errors.New("record with extra field")
+	ErrNotContainer = errors.New("scalar where container was expected")
+	ErrNotScalar    = errors.New("container where scaler was expected")
```
Comment (aswan) [inline pkg/zng/record.go]: typo here (scaler)

Comment (henridf) [reply pkg/zng/record.go]: fixed in #172

**Pull request merged**